### ISSUE TITLE
autograd for to_sparse(dim)

### ISSUE THREAD

**Issue by LiuXiaoxuanPKU** [pytorch/pytorch#46720]
## 🚀 Feature
autograd for to_sparse(dim)

## Motivation
This is related to another GitHub [issue](https://github.com/pytorch/pytorch/issues/18111), that issue is fixed by the [commit](https://github.com/pytorch/pytorch/pull/20458). However, it seems it only supports non-parameter version. I want to have autograd for to_sparse(dim). 

To be more concrete, the following (case 1) is supported
```
>>> torch.randn(2, 3).requires_grad_(True).to_sparse()
tensor(indices=tensor([[0, 0, 0, 1, 1, 1],
                       [0, 1, 2, 0, 1, 2]]),
       values=tensor([ 0.3336,  0.7028, -1.0183, -0.3018, -0.6059, -1.0752]),
       size=(2, 3), nnz=6, layout=torch.sparse_coo, grad_fn=<ToSparseBackward>)
```
 the following (case 2) is not supported
```
>>> torch.randn(2, 3).requires_grad_(True).to_sparse(1)
tensor(indices=tensor([[0, 1]]),
       values=tensor([[ 1.1035,  2.0671,  0.5015],
                      [ 0.0975,  0.3090, -1.1393]]),
       size=(2, 3), nnz=2, layout=torch.sparse_coo, grad_fn=<NotImplemented>)
```

Is it possible to support case 2?

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @aocsa @mruberry

**Comment by ezyang** [comment]
This seems reasonable to support.

**Comment by CurryYuan** [comment]
Any progress here?

**Comment by LiuXiaoxuanPKU** [comment]
The fix might not be the best one, but it works for me
https://github.com/pytorch/pytorch/pull/46825

> Any progress here?
